README.md
# uPydevice

[![PyPI version](https://badge.fury.io/py/upydevice.svg)](https://badge.fury.io/py/upydevice)[![PyPI license](https://img.shields.io/pypi/l/ansicolortags.svg)](https://pypi.python.org/pypi/ansicolortags/)

Python library to interface with MicroPython devices through REPL*:

- Websockets/WebREPL (WiFi)
- BleREPL (Bluetooth Low Energy)
- Serial REPL (USB)

\*  *Device REPL must be accesible.*

Upydevice allows to integrate device interaction from simple scripts to automated services,  test suites, command line interfaces or even GUI applications.

### Install

`pip install upydevice`  or `pip install --upgrade upydevice`

#### Example usage:

### DEVICE: `Device`

```python
>>> from upydevice import Device
```

This will return a Device based on address (first argument) type.

e.g.

### Serial (USB)

```python
>>> esp32 = Device('/dev/cu.usbserial-016418E3', init=True)
>>> esp32
SerialDevice @ /dev/cu.usbserial-016418E3, Type: esp32, Class: SerialDevice
Firmware: MicroPython v1.19.1-285-gc4e3ed964-dirty on 2022-08-12; ESP32 module with ESP32
CP2104 USB to UART Bridge Controller, Manufacturer: Silicon Labs
(MAC: 30:ae:a4:23:35:64)
```

### WiFi (WebSockets/WebREPL)

This requires [WebREPL](http://docs.micropython.org/en/latest/esp8266/tutorial/repl.html#webrepl-a-prompt-over-wifi) to be enabled in the device.

```python
>>> esp32 = Device('192.168.1.53', 'mypass', init=True)
>>> esp32
WebSocketDevice @ ws://192.168.1.53:8266, Type: esp32, Class: WebSocketDevice
Firmware: MicroPython v1.19.1-285-gc4e3ed964-dirty on 2022-08-12; ESP32 module with ESP32
(MAC: 30:ae:a4:23:35:64, Host Name: espdev, RSSI: -45 dBm)
```

If device hostname name is set e.g. (in device `main.py`)

```python
....
wlan = network.WLAN(network.STA_IF)
wlan.active(True)
wlan.config(hostname='espdev')
wlan.connect('your-ssid', 'your-key')
...
```

Now the device can be connected with mDNS name.

```python
>>> esp32 = Device('espdev.local', 'mypass', init=True)
>>> esp32
WebSocketDevice @ ws://192.168.1.53:8266, Type: esp32, Class: WebSocketDevice
Firmware: MicroPython v1.19.1-285-gc4e3ed964-dirty on 2022-08-12; ESP32 module with ESP32
(MAC: 30:ae:a4:23:35:64, Host Name: espdev, RSSI: -45 dBm)
```

### BLE (BleREPL)

This requires [BleREPL](https://upydev.readthedocs.io/en/latest/gettingstarted.html) to be enabled in the device. (This is still experimental and performance may be platform and device dependent)

```python
>>> esp32 = Device('00FEFE2D-5983-4D6C-9679-01F732CBA9D9', init=True)
>>> esp32
BleDevice @ 00FEFE2D-5983-4D6C-9679-01F732CBA9D9, Type: esp32 , Class: BleDevice
Firmware: MicroPython v1.18-128-g2ea21abae-dirty on 2022-02-19; 4MB/OTA BLE module with ESP32
(MAC: ec:94:cb:54:8e:14, Local Name: espble, RSSI: -38 dBm)
```

Now the device can recive commands and send back the output. e.g.

```python
>>> esp32.wr_cmd('led.on()')
>>> esp32.wr_cmd("os.listdir()")
['boot.py', 'webrepl_cfg.py', 'main.py'] # this output is stored in [upydevice].output

>>> esp32.output
['boot.py', 'webrepl_cfg.py', 'main.py']
>>>> esp32.wr_cmd('x = [1,2,3];my_var = len(x);print(my_var)')
3
# Soft Reset:
>>> esp32.reset()
    Rebooting device...
    Done!
```

### Testing devices with Pytest:

Under `test` directory there are example tests to run with devices. This allows to test MicroPython code in devices interactively, e.g. button press, screen swipes, sensor calibration, actuators, servo/stepper/dc motors ...
e.g.

```python
$ pytest test_esp_serial.py -s

=========================================== test session starts ===========================================
platform darwin -- Python 3.7.9, pytest-7.1.2, pluggy-1.0.0
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /Users/carlosgilgonzalez/Desktop/IBM_PROJECTS/MICROPYTHON/TOOLS/upydevice_.nosync/test, configfile: pytest.ini
plugins: benchmark-3.4.1
collected 7 items

test_esp_serial.py::test_devname PASSED                                                             [ 14%]
test_esp_serial.py::test_platform
---------------------------------------------- live log call ----------------------------------------------
00:13:26 [pytest] [sdev] [ESP32] : Running SerialDevice test...
00:13:26 [pytest] [sdev] [ESP32] : DEV PLATFORM: esp32
00:13:26 [pytest] [sdev] [ESP32] : DEV PLATFORM TEST: [✔]
PASSED                                                                                              [ 28%]
test_esp_serial.py::test_blink_led
---------------------------------------------- live log call ----------------------------------------------
00:13:28 [pytest] [sdev] [ESP32] : BLINK LED TEST: [✔]
PASSED                                                                                              [ 42%]
test_esp_serial.py::test_run_script
---------------------------------------------- live log call ----------------------------------------------
00:13:28 [pytest] [sdev] [ESP32] : RUN SCRIPT TEST: test_code.py
00:13:31 [pytest] [sdev] [ESP32] : RUN SCRIPT TEST: [✔]
PASSED                                                                                              [ 57%]
test_esp_serial.py::test_raise_device_exception
---------------------------------------------- live log call ----------------------------------------------
00:13:31 [pytest] [sdev] [ESP32] : DEVICE EXCEPTION TEST: b = 1/0
00:13:31 [pytest] [sdev] [ESP32] : DEVICE EXCEPTION TEST: [✔]
PASSED                                                                                              [ 71%]
test_esp_serial.py::test_reset
---------------------------------------------- live log call ----------------------------------------------
00:13:31 [pytest] [sdev] [ESP32] : DEVICE RESET TEST
00:13:32 [pytest] [sdev] [ESP32] : DEVICE RESET TEST: [✔]
PASSED                                                                                              [ 85%]
test_esp_serial.py::test_disconnect
---------------------------------------------- live log call ----------------------------------------------
00:13:32 [pytest] [sdev] [ESP32] : DEVICE DISCONNECT TEST
00:13:32 [pytest] [sdev] [ESP32] : DEVICE DISCONNECT TEST: [✔]
PASSED                                                                                              [100%]

============================================ 7 passed in 6.74s ============================================
```

### Made with upydevice:

- [upydev](https://github.com/Carglglz/upydev)
- [Jupyter_upydevice_kernel](https://github.com/Carglglz/jupyter_upydevice_kernel)

## [Examples (scripts, GUI ...)](https://github.com/Carglglz/upydevice/tree/develop/examples)


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)

## [0.3.9] Unreleased Github Repo [develop]
### Fix
- Drop self-signed certificates in CA path, fix SSL verification error.
## [0.3.8] - 2022-08-29
### Added
- Device `raise_traceback` method to catch Device Exception after follow mode
## [0.3.7] - 2022-08-23
### Fix
- paste_buff method in BleDevice
- fix data feeding pipe missing blank lines in follow callback for BleDevice
- deprecate self-signed certs in favor of ROOT CA signed certs for WebSocketDevices using SSL/TLS
### Added
- @device.code decorator for Device classes to declare and call functions in devices
- @device.code_follow for Device classes to declare and call functions in devices whose output need to be followed.
- @devicegroup.code for DeviceGroup class to declare and call functions in group of devices
– @devicegroup.code_follow for Device classes to declare and call functions in a group of devices whose output need to be followed.
- decorator examples
## [0.3.6] - 2022-03-25
### Fix
- fixed missing dependencies in `setup.py` (@BradenM)
## [0.3.5] - 2022-03-15
### Fix
- reassign ip to hostname if hostname is not None for disconnect method
## [0.3.4] - 2022-03-04
### Fix
- `WebSocketDevice` double unwanted `super().__init__` bug.
- Load time in MacOS caused by bleak/corebluetooth, only if BleDevice
- Allow mdns `.local` name in WebSecureREPL, (ssl checkhostname / auth server.)
- `WebSocketDevice` receive frame errors
- `SerialDevice` follow command parsing errors
- `BleDevice` follow command parsing errors
- `is_reachable` method for `WebSocketDevice` if ip is mdns .local name
- pipe option for `BleDevice` `wr_cmd` method.
### Added
- Indicate websocket port in ip address as `ip:port` format
- Allow `password:passphrase` format for `WebSocketDevice` password that sets ssl/auth
to `True` if present.
- `address` property for all device types.
- `flush_conn` for `BleDevice` for compatibility
## [0.3.3] 2021-12-16
### Added
- `Host Name` in `__repr__` command device info
## [0.3.2] 2021-10-24
### Added
- mDNS `.local`/`dhcp_hostname` compatibility, so device configuration works across networks
### Fix
- `check -i`, `info` commands in `-apmd`, if connected to AP of the device.
## [0.3.1] - 2021-09-09
### Fix
- KeyboardInterrupt in AsyncBleDevice for jupyter_upydevice_kernel
## [0.3.0] - 2021-08-12
#### Bump to version 0.3.0
### Added
- Tests with `pytest` tested with esp32 (websocket, serial, ble), esp8266(websocket, serial), pyboard (serial) and nrf52832 (colmi p8 with wasp-os) (ble)
- `Device` function that returns Device class based on address type
- Pytest with devices
- `__repr__` that returns device info
- Raise `DeviceException` on raised Exception in device
- Raise `DeviceNotFound` after connection attempt fail or if device is not reachable.
- `ble_scan`, `net_scan`, `serial_scan`, to scan for devices
### Fix
- Base code refactoring
## [0.2.3] - 2020-05-25
### Added
- `BLE_DEVICE` for bluetooth low energy devices
### Fix
- Serial device `is_reachable` method fix
## [0.2.2] - 2020-02-24
### Fix
- WS_DEVICE cmd_nb fix for cmd method
## [0.2.1] - 2020-02-17
### Fix
- WS_DEVICE now works with esp8266 too.
## [0.2.0] - 2020-02-16
### Added
- NEW WS_DEVICE and SERIAL_DEVICE work in progress
## [0.1.9] - 2020-02-08
### Added
- BASE_WS_DEVICE compatibility with WebSecureREPL

## [0.1.8] - 2020-02-02
### Added
- BASE_SERIAL_DEVICE and BASE_WS_DEVICE work in progress
## [0.1.7] - 2020-01-19
### Fix
- W_UPYDEVICE kbi command can filter MicroPython banner info now
## [0.1.6] - 2020-01-10
### Added
- Reset option in "init" S_UPYDEVICE for upydev 'sh_srepl' compatibility
## [0.1.5] - 2020-01-07
### Added
- phantom_wr module like phantom but faster
### Fix
- KeyboardInterrupt handling in wr_cmd method of W_UPYDEVICE, S_UPYDEVICE, PYBOARD
## [0.1.4] - 2019-12-24
### Added
- New methods 'open_wconn', 'wr_cmd' and 'close_wconn' to open a wrepl session
and send commands (this will open a WebREPL connection and wont close it until
close_wconn is called, which means the other commands methods won't work) This
will drop the delay between sending a command and executing that command on the
device
- Methods 'open_wconn', 'wr_cmd' and 'close_wconn' for serial devices too.
(With serial REPL instead of WebREPL)
### Fix
- uparser_dec improved
## [0.1.3] - 2019-11-23
### Added
- Methods to UOS phantom class (mkdir, rmdir, getcwd, chdir, remove)
- Phantom pyb_Servo pyboard class
- New sensors IRQ phantom classes
## [0.1.2] - 2019-11-12
### Added
- New decorator (upy_cmd_c_r_in_callback) to be able to pass self methods (functions) as keywords arguments (for callbacks)
- New decorator that allows non-blocking function calls (upy_cmd_c_r_nb)
- New "IRQ_util.py" upy script (U_IRQ_MG class) + "Phantom" python IRQ_MG class
  This allow to manage/set interrupts, and receive and interrupt callback on python3 (see Phantom class Docs)
- New W_UPYDEVICE method 'is_reachable' (pings the device to see if it is
  reachable)
- New decorator (upy_cmd_c_r_nb_in_callback) to be able to pass self methods (functions) as keywords arguments (for callbacks) in non-blocking style
(good for timer interrupts)
- New "STREAMER_util.py" upy script (U_STREAMER class) as a "super" class to
  make streaming sensor data in real time easier (up to ~900 Hz of sampling frequency tested) + "Phantom" python STREAMER class (see Phantom class Docs)
- New phantom sensors classes (ADS1115, BME280, LSM9DS1)
- New phantom sensors STREAMER classes (IMU_STREAMER, BME_STREAMER,
  ADS_STREAMER)+ upydevice_utils scripts (IMU_util.py, ADS_util.py,
  BME_util.py) to test U_STREAMER and U_IRQ_MG classes (more about this in docs
  and examples)
## [0.1.1] - 2019-11-02
### Added

upydevice/protocol.py
"""
Websockets protocol
"""

import logging
import re
import struct
import random
import socket
from collections import namedtuple

LOGGER = logging.getLogger(__name__)
LOGGER.setLevel(logging.INFO)

# Opcodes
OP_CONT = 0x0
OP_TEXT = 0x1
OP_BYTES = 0x2
OP_CLOSE = 0x8
OP_PING = 0x9
OP_PONG = 0xa

# Close codes
CLOSE_OK = 1000
CLOSE_GOING_AWAY = 1001
CLOSE_PROTOCOL_ERROR = 1002
CLOSE_DATA_NOT_SUPPORTED = 1003
CLOSE_BAD_DATA = 1007
CLOSE_POLICY_VIOLATION = 1008
CLOSE_TOO_BIG = 1009
CLOSE_MISSING_EXTN = 1010
CLOSE_BAD_CONDITION = 1011

URL_RE = re.compile(r'(wss|ws)://([A-Za-z0-9-\.]+)(?:\:([0-9]+))?(/.+)?')
URI = namedtuple('URI', ('protocol', 'hostname', 'port', 'path'))

class NoDataException(Exception):
    pass

class ConnectionClosed(Exception):
    pass

def urlparse(uri):
    """Parse ws:// URLs"""
    match = URL_RE.match(uri)
    if match:
        protocol = match.group(1)
        host = match.group(2)
        port = match.group(3)
        path = match.group(4)

        if protocol == 'wss':
            if port is None:
                port = 443
        elif protocol == 'ws':
            if port is None:
                port = 80
        else:
            raise ValueError('Scheme {} is invalid'.format(protocol))

        return URI(protocol, host, int(port), path)


class Websocket:
    """
    Basis of the Websocket protocol.

    This can probably be replaced with the C-based websocket module, but
    this one currently supports more options.
    """
    is_client = False

    def __init__(self, sock):
        self.sock = sock
        self.open = True

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc, tb):
        self.close()

    def settimeout(self, timeout):
        self.sock.settimeout(timeout)

    def read_frame(self, max_size=None):
        """
        Read a frame from the socket.
        See https://tools.ietf.org/html/rfc6455#section-5.2 for the details.
        """
        two_bytes = b''
        # Frame header
        two_bytes += self.sock.recv(2)
        while len(two_bytes) < 2:
            two_bytes += self.sock.recv(1)

        if not two_bytes:
            raise NoDataException

        byte1, byte2 = struct.unpack('!BB', two_bytes)

        # Byte 1: FIN(1) _(1) _(1) _(1) OPCODE(4)
        fin = bool(byte1 & 0x80)
        opcode = byte1 & 0x0f

        # Byte 2: MASK(1) LENGTH(7)
        mask = bool(byte2 & (1 << 7))
        length = byte2 & 0x7f

        if length == 126:  # Magic number, length header is 2 bytes
            length, = struct.unpack('!H', self.sock.read(2))
        elif length == 127:  # Magic number, length header is 8 bytes
            length, = struct.unpack('!Q', self.sock.read(8))

        if mask:  # Mask is 4 bytes
            mask_bits = self.sock.recv(4)

        try:
            data = self.sock.recv(length)
        except MemoryError:
            # We can't receive this many bytes, close the socket
            if __debug__: LOGGER.debug("Frame of length %s too big. Closing",
                                       length)
            self.close(code=CLOSE_TOO_BIG)
            return True, OP_CLOSE, None

        if mask:
            data = bytes(b ^ mask_bits[i % 4]
                         for i, b in enumerate(data))

        return fin, opcode, data

    def write_frame(self, opcode, data=b''):
        """
        Write a frame to the socket.
        See https://tools.ietf.org/html/rfc6455#section-5.2 for the details.
        """
        fin = True
        mask = self.is_client  # messages sent by client are masked

        length = len(data)

        # Frame header
        # Byte 1: FIN(1) _(1) _(1) _(1) OPCODE(4)
        byte1 = 0x80 if fin else 0
        byte1 |= opcode

        # Byte 2: MASK(1) LENGTH(7)
        byte2 = 0x80 if mask else 0

        if length < 126:  # 126 is magic value to use 2-byte length header
            byte2 |= length
            self.sock.send(struct.pack('!BB', byte1, byte2))

        elif length < (1 << 16):  # Length fits in 2-bytes
            byte2 |= 126  # Magic code
            self.sock.send(struct.pack('!BBH', byte1, byte2, length))

        elif length < (1 << 64):
            byte2 |= 127  # Magic code
            self.sock.send(struct.pack('!BBQ', byte1, byte2, length))

        else:
            raise ValueError()

        if mask:  # Mask is 4 bytes
            mask_bits = struct.pack('!I', random.getrandbits(32))
            self.sock.send(mask_bits)

            data = bytes(b ^ mask_bits[i % 4]
                         for i, b in enumerate(data))

        self.sock.send(data)

    def recv(self):
        """
        Receive data from the websocket.

        This is slightly different from 'websockets' in that it doesn't
        fire off a routine to process frames and put the data in a queue.
        If you don't call recv() sufficiently often you won't process control
        frames.
        """
        assert self.open

        while self.open:
            try:
                fin, opcode, data = self.read_frame()
            except NoDataException:
                return ''
            except ValueError:
                LOGGER.debug("Failed to read frame. Socket dead.")
                self._close()
                raise ConnectionClosed()

            if not fin:
                raise NotImplementedError()

            if opcode == OP_TEXT:
                return data.decode('utf-8')
            elif opcode == OP_BYTES:
                return data
            elif opcode == OP_CLOSE:
                self._close()
                return
            elif opcode == OP_PONG:
                # Ignore this frame, keep waiting for a data frame
                continue
            elif opcode == OP_PING:
                # We need to send a pong frame
                if __debug__: LOGGER.debug("Sending PONG")
                self.write_frame(OP_PONG, data)
                # And then wait to receive
                continue
            elif opcode == OP_CONT:
                # This is a continuation of a previous frame
                raise NotImplementedError(opcode)
            else:
                raise ValueError(opcode)

    def send(self, buf):
        """Send data to the websocket."""

        assert self.open

        if isinstance(buf, str):
            opcode = OP_TEXT
            buf = buf.encode('utf-8')
        elif isinstance(buf, bytes):
            opcode = OP_BYTES
        else:
            raise TypeError()

        self.write_frame(opcode, buf)

    def close(self, code=CLOSE_OK, reason=''):
        """Close the websocket."""
        if not self.open:
            return

        buf = struct.pack('!H', code) + reason.encode('utf-8')

        self.write_frame(OP_CLOSE, buf)
        self._close()

    def _close(self):
        if __debug__: LOGGER.debug("Connection closed")
        self.open = False
        self.sock.close()
